### PR TITLE
WIP: Fragments

### PR DESCRIPTION
**Pull request by QuiCM**
*THIS PR IS PRIMARILY FOR DISCUSSION PURPOSES & REVIEWS DURING THE DEVELOPMENT OF THE FRAGMENTS BRANCH. **DO NOT** MERGE WITHOUT MY APPROVAL*

To contribute to this PR, push commits to the Fragments branch (**with approval from @hakusaro**).

### CODE REVIEW

File: TShockAPI/Bouncer.cs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+using System.IO.Streams;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Terraria.ID;
+using TShockAPI.Net;
+
+namespace TShockAPI
+{
+	/// <summary>Bouncer - The TShock anti-hack and build guardian system</summary>
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Comments do not need to (should not, imo) include the name of the object being commented.
A class level summary should provide high level details on what the class does (but more detailed than "anti-hack and build guardian system").

File: TShockAPI/Bouncer.cs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+	public class Bouncer
+	{
+		/// <summary>Bouncer - Constructor call initializes Bouncer & related functionality.</summary>
+		/// <returns>A new Bouncer.</returns>
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Returns on a constructor is a bit redundant. If this is to remain, add more details about what setup the returned instance has

File: TShockAPI/Bouncer.cs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+		public Bouncer(TerrariaPlugin pluginInstance)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Do we need a public constructor? Why does it take a `TerrariaPlugin` reference?

File: TShockAPI/Bouncer.cs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Terraria.ID;
+using TShockAPI.Net;
+
+namespace TShockAPI
+{
+	/// <summary>Bouncer - The TShock anti-hack and build guardian system</summary>
+	public class Bouncer
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Is Bouncer intended to be extensible? If not this should be `sealed`.
If it is intended to be extensible we should provide an abstract base class or an interface from which this implementation inherits and future implementations can use

File: TShockAPI/Bouncer.cs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+		/// <returns>A new Bouncer.</returns>
+		public Bouncer(TerrariaPlugin pluginInstance)
+		{
+			// Setup hooks
+
+			GetDataHandlers.SendTileSquare.Register(OnSendTileSquare);
+			GetDataHandlers.HealOtherPlayer.Register(OnHealOtherPlayer);
+		}
+
+		/// <summary>OnHealOtherPlayer - The handler for the HealOther events in Bouncer</summary>
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Again, summaries don't need the method name either. Description could be more descriptive, especially for the parameters (`<param name />`s below)

File: TShockAPI/Bouncer.cs [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+		/// <param name="sender">sender</param>
+		/// <param name="args">args</param>
+		internal void OnHealOtherPlayer(object sender, GetDataHandlers.HealOtherPlayerEventArgs args)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: If Bouncer is intended to be extensible then this should be `public virtual`, or, assuming a base interface provides this method and bouncer inherits, `public override`

File: TShockAPI/Bouncer.cs [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+			short amount = args.Amount;
+			byte plr = args.TargetPlayerIndex;
+
+			if (amount <= 0 || Main.player[plr] == null || !Main.player[plr].active)
+			{
+				args.Handled = true;
+				return;
+			}
+
+			if (amount > TShock.Config.MaxDamage * 0.2)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Why are we checking max damage * 0.2 for healing others?

File: TShockAPI/Bouncer.cs [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+			}
+
+			if (args.Player.HealOtherThreshold > TShock.Config.HealOtherThreshold)
+			{
+				args.Player.Disable("Reached HealOtherPlayer threshold.", DisableFlags.WriteToLogAndConsole);
+				args.Handled = true;
+				return;
+			}
+
+			if (TShock.CheckIgnores(args.Player) || (DateTime.UtcNow - args.Player.LastThreat).TotalMilliseconds < 5000)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: A comment here would be good.
I assume this rejects the packet if the player is
* 'ignored' due to other hack detection
* has been reprimanded for some other potential hack attempt in the last 5 seconds

File: TShockAPI/Bouncer.cs [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+			args.Player.HealOtherThreshold++;
+			args.Handled = false;
+			return;
+		}
+
+		/// <summary>OnSendTileSquare - The handler for SendTileSquare events in Bouncer</summary>
+		/// <param name="sender">sender</param>
+		/// <param name="args">args</param>
+		internal void OnSendTileSquare(object sender, GetDataHandlers.SendTileSquareEventArgs args)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Comments and design again, see above review comments. I'll stop marking these from here-on because the feedback is the same

File: TShockAPI/Bouncer.cs [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+			int tileX = args.TileX;
+			int tileY = args.TileY;
+
+			if (args.Player.HasPermission(Permissions.allowclientsideworldedit))
+			{
+				args.Handled = false;
+				return;
+			}
+
+			if (size > 5)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: We discussed this - 5 is a magic number, we should probably figure out exactly why we use 5.
IIRC it's because 5 means a 5x5 square which is normal for a tile square, and anything bigger is a non-vanilla tile modification attempt

File: TShockAPI/Bouncer.cs [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+			{
+				args.Handled = true;
+				return;
+			}
+
+			if ((DateTime.UtcNow - args.Player.LastThreat).TotalMilliseconds < 5000)
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: Honestly we should just convert this into a method. Something like `bool IsWarned() => (DateTime.UtcNow - LastThread).TotalMilliseconds < 5000;` in `TSPlayer`

**Comment by Ijwu** [comment]
Will do an official review before merge to ensure I get final code.

So far, it looks good. I appreciate the attempt to compartmentalize various functionality. In the long run this will be a great improvement. 👍

File: TShockAPI/Bouncer.cs [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+			{
+				args.Player.SendTileSquare(tileX, tileY, size);
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: A comment on what SendTileSquare does would be good.
Something along the lines of "Undoes the action for the client that performed it. E.G., if we reject a tile break this sends the server's state of the tile so the client doesn't see a broken tile"

File: TShockAPI/Bouncer.cs [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+			}
+
+			if (TShock.CheckIgnores(args.Player))
+			{
+				args.Player.SendTileSquare(tileX, tileY, size);
+				args.Handled = true;
+				return;
+			}
+
+			try
```
Comment (QuiCM) [inline TShockAPI/Bouncer.cs]: This is a big logic block that's pretty arcane to look at.
Maybe add a short summary above the `try` statement that explains what happens in the block

File: TShockAPI/GetDataHandlers.cs [hunk 14]
```diff
@@ -532,19 +532,64 @@ public static bool TSCheckNoclip(Vector2 Position, int Width, int Height)
 			return false;
 		}
 
+		/// <summary>HandledEventArgs - The event args object for the HealOtherPlayer event</summary>
```
Comment (QuiCM) [inline TShockAPI/GetDataHandlers.cs]: Comments again

File: TShockAPI/TShock.cs [hunk 15]
```diff
@@ -378,6 +378,8 @@ public override void Initialize()
 				Log.ConsoleInfo("Welcome to TShock for Terraria!");
 				Log.ConsoleInfo("TShock comes with no warranty & is free software.");
 				Log.ConsoleInfo("You can modify & distribute it under the terms of the GNU GPLv3.");
+
+				Bouncer b = new Bouncer(this);
```
Comment (QuiCM) [inline TShockAPI/TShock.cs]: This is a local variable. Will go out of scope and get GCed, won't it?

If dotnet magic keeps it alive, it's inaccessible externally and from the rest of the codebase

File: TShockAPI/Utils.cs [hunk 16]
```diff
@@ -1427,5 +1427,66 @@ private void OutputElementsForDump(StringBuilder buffer, List<object[]> elements
 				}
 			}
 		}
+
+		/// <summary>FixChestStacks - Verifies that each stack in each chest is valid and not over the max stack count.</summary>
```
Comment (QuiCM) [inline TShockAPI/Utils.cs]: Comment style

File: TShockAPI/Bouncer.cs (hunk 6 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Not.

File: TShockAPI/Bouncer.cs (hunk 3 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Well, it might need one eventually (not done yet).

File: TShockAPI/Bouncer.cs (hunk 4 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: I don't want to hear the word "interface" without good justification ever again.

File: TShockAPI/Bouncer.cs (hunk 1 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: > Comments do not need to (should not, imo) include the name of the object being commented.

Got it.

> A class level summary should provide high level details on what the class does (but more detailed than "anti-hack and build guardian system").

Given that I don't know how much of our code is anti-hack, it would be quite stupid to predict the future and say what it does without knowing what it does.

File: TShockAPI/Bouncer.cs (hunk 7 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Great question.

cc @Pryaxis/tshock does anyone know why?

File: TShockAPI/Bouncer.cs (hunk 9 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Thank god.

File: TShockAPI/Bouncer.cs (hunk 11 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Noted for future use (I have objectives that are more important than this though).

File: TShockAPI/TShock.cs (hunk 15 above)
Comment (hakusaro) [reply TShockAPI/TShock.cs]: screen shot 2017-12-06 at 16 49 23

Comment (hakusaro) [reply TShockAPI/TShock.cs]: I wanted code compiling not being correct for final implementation.

Comment (Ijwu) [reply TShockAPI/TShock.cs]: Not for as long as it's hooks are registered, I believe.

The `Bouncer` constructor does that.

**Comment by hakusaro** [comment]
The fragments branch is a pet project by me. If you want to contribute, send me a message on Discord and swear under oath that you will not abandon me ever.

File: TShockAPI/TShock.cs (hunk 15 above)
Comment (hakusaro) [reply TShockAPI/TShock.cs]: @Ijwu it's still not remotely close to done because there's no destructor and no other stuff (this is not final obviously).

Comment (Ijwu) [reply TShockAPI/TShock.cs]: Understood. Just commenting on current state. Thanks! :)

File: TShockAPI/Bouncer.cs (hunk 4 above)
Comment (QuiCM) [reply TShockAPI/Bouncer.cs]: In that case I assume this class is meant for internal use only, and should not be extensible by external parties (eg plugins).
`public sealed`, or even better, `internal sealed` markings should be applied

**Comment by hakusaro** [comment]
I'm gonna update the topic for this issue eventually. Let me summarize what this branch is going to be in brief in the mean time.

This is going to be moving everything that TShock does out into subsystems based on logical groups. For example, the anti-hack code is currently the primary resident of GetDataHandlers, but it has utility functions spread all throughout TShock. These methods, like ```CheckIgnores``` just sit around and have very little descriptive purpose as to what they do. I'm going through and moving all of the code that calls these methods into different ```fragments``` of the original code.

At this point, I expect the first merge of this branch will happen once the anti-hack code is moved into its new home: ```Bouncer```. This project will likely take a non-insignificant amount of time, so I don't know exactly when I'll get there. Hopefully things will pick up as I start to isolate the anti-hack system.

The anti-hack system is decrepit and needs updating and revisiting for 2017. The problem is that updating it right now is difficult because reasoning about what it does is basically impossible. We need to have a hard look at it, but it's actually pretty low on my priority list. I want to get the ban system rewritten first, and that requires getting new database stuff done.

But all of that requires that we have logical groupings of subsystems. Essentially, I'm going the _gizmo_ route except not trying to re-implement all of TShock in one-go or trying to split all the subsystem stuff out in one shot.

The current basis for how I'm acting is by looking at things that call ```CheckIgnores``` and moving them into the Bouncer subsystem. Eventually, that's going to be extended to other methods in the main class and in the utils file. Any anti-hack/anti-cheat stuff should end up in ```Bouncer``` eventually.

The subsystem after ```Bouncer``` for isolation is SSC. Once that's done, I'm going to isolate smaller functional groups into a subsystem that will encompass TShock's minor tweaks. These are like perpetual events, random checks, and the initialize block. My goal is to get ```TShock.cs``` down to as few lines as possible of initialization code of some core subsystems. Those core subsystems will be ordered logically and easily changed by future code maintainers. It should be really stupid obvious just based on file structure what systems go where and how they can be modified.

That's my goal.

File: TShockAPI/Bouncer.cs (hunk 7 above)
Comment (bartico6) [reply TShockAPI/Bouncer.cs]: Because heal other player only happens when you are using the spectre armor with the hood, and the healing you can do with that is 20% of your damage.

File: TShockAPI/Bouncer.cs [hunk 17]
```diff
@@ -40,11 +41,138 @@ internal Bouncer(TerrariaPlugin pluginInstance)
...
+
+			if (!TShock.Utils.TilePlacementValid(x, y))
+			{
+				args.Handled = true;
+				return;
+			}
+
+			if (args.Player.Dead && TShock.Config.PreventDeadModification)
+			{
+				args.Handled = true;
```
Comment (bartico6) [inline TShockAPI/Bouncer.cs]: This should also send the tiles back to the user. If the user is dead and modifying tiles, this causes desync - the client will see a tunnel that nobody else sees, and appear to be noclipping. A workaround is to also send the tile back *before* 'handling' it. Just add `args.Player.SendTileSquare(x, y, 1);` 👍

File: TShockAPI/Bouncer.cs [hunk 18]
```diff
@@ -0,0 +1,803 @@
...
+					if (TShock.Itembans.ItemIsBanned(EnglishLanguage.GetItemNameById(selectedItem.netID), args.Player) || editData >= (action == EditAction.PlaceTile ? Main.maxTileSets : Main.maxWallTypes))
+					{
+						args.Player.SendTileSquare(tileX, tileY, 4);
+						args.Handled = true;
+						return;
+					}
+					if (action == EditAction.PlaceTile && (editData == 29 || editData == 97) && Main.ServerSideCharacter)
+					{
+						args.Player.SendErrorMessage("You cannot place this tile because server side characters are enabled.");
+						args.Player.SendTileSquare(tileX, tileY, 3);
```
Comment (mistzzt) [inline TShockAPI/Bouncer.cs]: If it is possible to use a constant to eliminate magic numbers here?

Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: Sure. Do you know what those magic numbers are?

Comment (mistzzt) [reply TShockAPI/Bouncer.cs]: I think different numbers should represent the area of some tiles. 
1 for one tile, 4 for some large objects.
Here, the `editData` is the tile id for piggy bank and safes; Idk why the size of tile square is 3.

Comment (QuiCM) [reply TShockAPI/Bouncer.cs]: We can replace magic numbers with the related `TileID.TileType` constants

Comment (QuiCM) [reply TShockAPI/Bouncer.cs]: Also I imagine it's 3 because that's a 3x3 square (safe is a 2x2, piggy is a 2x1 irc) and we wanted to be safe

File: TShockAPI/Bouncer.cs (hunk 12 above)
Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: @QuiCM documented in eed7b3fa.

File: TShockAPI/Bouncer.cs [hunk 19]
```diff
@@ -201,12 +201,14 @@ internal void OnPlaceObject(object sender, GetDataHandlers.PlaceObjectEventArgs
 
 			if (!TShock.Utils.TilePlacementValid(x, y))
 			{
+				args.Player.SendTileSquare(tileX, tileY, 4);
```
Comment (bartico6) [inline TShockAPI/Bouncer.cs]: Hate to be such a pain about this, but I'm almost [99%] certain that sending a 1-big box is enough - that's what I do on my server and it has allowed me to completely fix the issue of clientside tunneling by modifying tiles while dead. Since only one tile was modified, you should resend just that one tile rather than four - why waste bandwidth :-)

File: TShockAPI/Bouncer.cs [hunk 20]
```diff
@@ -201,12 +201,14 @@ internal void OnPlaceObject(object sender, GetDataHandlers.PlaceObjectEventArgs
...
 				args.Handled = true;
 				return;
 			}
 
 			if (args.Player.Dead && TShock.Config.PreventDeadModification)
 			{
+				args.Player.SendTileSquare(tileX, tileY, 4);
```
Comment (bartico6) [inline TShockAPI/Bouncer.cs]: Same as above

File: TShockAPI/Bouncer.cs [hunk 21]
```diff
@@ -201,14 +201,14 @@ internal void OnPlaceObject(object sender, GetDataHandlers.PlaceObjectEventArgs
 
 			if (!TShock.Utils.TilePlacementValid(x, y))
 			{
-				args.Player.SendTileSquare(tileX, tileY, 4);
+				args.Player.SendTileSquare(x, y, 4);
```
Comment (bartico6) [inline TShockAPI/Bouncer.cs]: Re-posting since it got insta-outdated.  
```Hate to be such a pain about this, but I'm almost [99%] certain that sending a 1-big box is enough - that's what I do on my server and it has allowed me to completely fix the issue of clientside tunneling by modifying tiles while dead. Since only one tile was modified, you should resend just that one tile rather than four - why waste bandwidth :-)```

File: TShockAPI/Bouncer.cs [hunk 22]
```diff
@@ -201,14 +201,14 @@ internal void OnPlaceObject(object sender, GetDataHandlers.PlaceObjectEventArgs
...
 				args.Handled = true;
 				return;
 			}
 
 			if (args.Player.Dead && TShock.Config.PreventDeadModification)
 			{
-				args.Player.SendTileSquare(tileX, tileY, 4);
+				args.Player.SendTileSquare(x, y, 4);
```
Comment (bartico6) [inline TShockAPI/Bouncer.cs]: Same here, as previously.

File: TShockAPI/GetDataHandlers.cs [hunk 23]
```diff
@@ -2034,6 +2037,9 @@ private static bool HandlePlayerUpdate(GetDataHandlerArgs args)
 		{
 			if (args.Player == null || args.TPlayer == null || args.Data == null)
 			{
+				// Is this really the best option?
+				// If we're getting a packet that doesn't have a player or a TPlayer or data...
+				// Should we really let it through?
 				return false;
```
Comment (hakusaro) [inline TShockAPI/GetDataHandlers.cs]: Looking for input on if this is the best return value here.

File: TShockAPI/GetDataHandlers.cs [hunk 24]
```diff
@@ -2154,6 +2087,7 @@ private static bool HandlePlayerUpdate(GetDataHandlerArgs args)
 				}
 			}
 
+			// Where we rebuild sync data for Terraria?
```
Comment (hakusaro) [inline TShockAPI/GetDataHandlers.cs]: I'm also wondering why this block of code does what it does. Does Terraria not handle this?

File: TShockAPI/GetDataHandlers.cs [hunk 25]
```diff
@@ -2134,6 +2066,7 @@ private static bool HandlePlayerUpdate(GetDataHandlerArgs args)
 					args.Player.SendErrorMessage("You cannot use {0} on this server. Your actions are being ignored.", itemName);
 				}
 
+				// Reimplementation of normal Terraria stuff?
```
Comment (hakusaro) [inline TShockAPI/GetDataHandlers.cs]: Why is this here too?

File: TShockAPI/GetDataHandlers.cs (hunk 23 above)
Comment (bartico6) [reply TShockAPI/GetDataHandlers.cs]: Packets that come from null players might be packets sent prematurely by unconnected clients, this should be filtered to avoid exceptions further down the execution chain

Comment (hakusaro) [reply TShockAPI/GetDataHandlers.cs]: @bartico6 yeah, but by returning _false_ we're letting the server process it. If we return _true_ the event is cancelled and doesn't get sent to the server.

File: TShockAPI/GetDataHandlers.cs (hunk 25 above)
Comment (bartico6) [reply TShockAPI/GetDataHandlers.cs]: I think this is to avoid a client just sending a "health update" with arbitrary values to set their maxhealth, so instead you'd manually detect events which increase max hp and replay them against the server character data

File: TShockAPI/GetDataHandlers.cs (hunk 23 above)
Comment (hakusaro) [reply TShockAPI/GetDataHandlers.cs]: I _think_ this should be _true_

Comment (bartico6) [reply TShockAPI/GetDataHandlers.cs]: If true means handled then yes, it should be true IMO.

Comment (hakusaro) [reply TShockAPI/GetDataHandlers.cs]: @bartico6 I'm waiting on @QuiCM to confirm before changing this. He's the one who implemented the check, and I don't want to change behavior without 100% assurance of what the intention was.

File: TShockAPI/Bouncer.cs [hunk 26]
```diff
@@ -51,6 +52,33 @@ internal Bouncer(TerrariaPlugin pluginInstance)
...
 
+		/// <summary>Handles ProjectileKill events for throttling & out of bounds projectiles.</summary>
+		/// <param name="sender">The object that triggered the event.</param>
+		/// <param name="args">The packet arguments that the event has.</param>
+		internal void OnProjectileKill(object sender, GetDataHandlers.ProjectileKillEventArgs args)
+		{
+			if (args.ProjectileIndex > Main.maxProjectiles || args.ProjectileIndex < 0)
+			{
+				// TODO: Should this be /true/ to stop the server from processing it?
+				args.Handled = false;
```
Comment (hakusaro) [inline TShockAPI/Bouncer.cs]: Should this be true to stop the server from processing it?

File: TShockAPI/GetDataHandlers.cs (hunk 23 above)
Comment (QuiCM) [reply TShockAPI/GetDataHandlers.cs]: I don't recall why I did false. I can't think of any reason why we shouldn't prevent propagation of a bad packet.

I'm 100% fine with this changing to true

File: TShockAPI/Bouncer.cs (hunk 26 above)
Comment (QuiCM) [reply TShockAPI/Bouncer.cs]: I wonder if Terraria has any special logic for projectile id = -1.
However this looks like inverted logic again (where someone assumed false meant it was handled)

Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: I whipped out my handy dandy blame tool. Originally implemented by Zido as a `return true`, changed to `return false` by Deathmax in a commit called "merge fail."

screen shot 2017-12-11 at 00 21 19

screen shot 2017-12-11 at 00 21 44

Mind if I change it to true, @QuiCM?

Comment (hakusaro) [reply TShockAPI/Bouncer.cs]: I checked the [merge fail](https://github.com/Pryaxis/TShock/commit/672d360e9dbe6c93997f4d17ac0f5e2c987db6ea) commit and I definitely don't know what to say about it.

**Comment by hakusaro** [comment]
Progress update: 15/22 `CheckIgnores` calls to go (31% complete). I currently expect this to be completed by _Friday (Saturday Australia)_.

**Comment by hakusaro** [comment]
There should be at least 424 more lines that can be chopped out of TShock's main class by the end of fragments (v1). Potentially a ton more if I move all of the stupid `OnSecondUpdate` stuff to Bouncer (if applicable).

**Comment by hakusaro** [comment]
50%, 11 more calls to go.

**Comment by hakusaro** [comment]
Currently, it looks like we need to have a subsystem for regions and a subsystem for item bans, broken out separately.
